src/preamble.js

```diff
@@ -1044,50 +1044,51 @@ function enlargeMemory() {
   // TOTAL_MEMORY is the current size of the actual array, and DYNAMICTOP is the new top.
 #if ASSERTIONS
   assert(HEAP32[DYNAMICTOP_PTR>>2] > TOTAL_MEMORY); // This function should only ever be called after the ceiling of the dynamic heap has already been bumped to exceed the current total size of the asm.js heap.
-  assert(TOTAL_MEMORY > 4); // So the loop below will not be infinite
 #endif
 
-  var OLD_TOTAL_MEMORY = TOTAL_MEMORY;
-
 #if EMSCRIPTEN_TRACING
   // Report old layout one last time
   _emscripten_trace_report_memory_layout();
 #endif
 
-  var LIMIT = Math.pow(2, 31); // 2GB is a practical maximum, as we use signed ints as pointers
-                               // and JS engines seem unhappy to give us 2GB arrays currently
-  if (HEAP32[DYNAMICTOP_PTR>>2] >= LIMIT) return false;
+  if (Module["usingWasm"]) {
+    var LIMIT = 2147483648-PAGE_SIZE; // In Wasm, we can do one 64KB page short of 2GB as theoretical maximum.
+  } else {
+    var LIMIT = 2147483648-16777216; // In asm.js, the theoretical maximum is 16MB short of 2GB.
```

Comment:
perhaps the asm.js limit could just be the same as wasm? not much difference between them, and would improve seamlessness of asm/wasm transition